### PR TITLE
feat: fix and create smart documents e2e tests

### PR DESCRIPTION
**Pull request by infonl-marcel**
fix and create smart documents e2e tests

Solves: PZ-4194

### CODE REVIEW

File: src/e2e/step-definitions/external-systems.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * SPDX-License-Identifier: EUPL-1.2+
+ */
+
+import { Then, When } from "@cucumber/cucumber";
+import { Page } from "@playwright/test";
+import { CustomWorld } from "../support/worlds/world";
+
+const ONE_MINUTE_IN_MS = 60_000;
+
+// Reference to smartdocuments page
```
Comment (edgarvonk) [inline src/e2e/step-definitions/external-systems.ts:12]: Comment does not really add any value I think so I would remove it. PS: what is the SmartDocuments page in the first place?

File: src/e2e/step-definitions/external-systems.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+);
+
+When(
+  "Employee {string} fills in the create document form",
+  { timeout: ONE_MINUTE_IN_MS },
+  async function (this: CustomWorld, user) {
+    const submitButton = this.page.locator("#opslaan_button");
+    await submitButton.waitFor({ state: "visible" });
+
+    // Verify that the button is initially disabled
```
Comment (edgarvonk) [inline src/e2e/step-definitions/external-systems.ts:34]: Same here. I would remove this comment. The problem with code comments is that often at some point in time they grow out of sync with the code so I would only add them if it really adds value. Ideally code should be self-explanatory. :-)

File: src/e2e/step-definitions/external-systems.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    await klaarButton.waitFor({ state: "visible" });
+    await klaarButton.click();
+  },
+);
+
+Then(
+  "Employee {string} should see a result page mentioning the document is created",
+  { timeout: ONE_MINUTE_IN_MS },
+  async function (this: CustomWorld, user) {
+    // Last things to test here: https://dimpact.atlassian.net/browse/PZ-4251
```
Comment (edgarvonk) [inline src/e2e/step-definitions/external-systems.ts:85]: We would rather not commit any commented out code. Git is not a good medium to manage TODO stuff. Better to keep this somewhere else (a separate Git branch, JIRA, or whatever) because else code can quickly become unmaintainable (nothing worse than old TODO's in code of which nobody knows the status..).

**Review by edgarvonk — CHANGES_REQUESTED** [review]

File: src/e2e/step-definitions/external-systems.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    await this.page.getByLabel("Sjabloon").last().click();
+    await this.page.getByRole("option", { name: "OpenZaakTest" }).click();
+
+    const autofillInputTitle = this.page.locator("#title_tekstfield");
+    await autofillInputTitle.click();
+    await autofillInputTitle.fill("Document Title Text");
+
+    // Now check if the button is enabled after those actions
+    await submitButton.isEnabled();
```
Comment (bas-info-nl) [inline src/e2e/step-definitions/external-systems.ts:52]: What happens if the button is not, and will never be, enabled? Does it fail with a timeout?

File: src/e2e/step-definitions/external-systems.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  },
+);
+
+When(
+  "Employee {string} submits the form to create the document should see SmartDocuments tab",
+  { timeout: ONE_MINUTE_IN_MS },
+  async function (this: CustomWorld, user) {
+    await this.page.click("#opslaan_button");
+
+    smartDocumentsPage = await this.page.waitForEvent("popup");
```
Comment (bas-info-nl) [inline src/e2e/step-definitions/external-systems.ts:62]: I agree with Edgar's comment that this name is not descriptive enough, I'll make a suggestion on the name in another comment.

File: src/e2e/step-definitions/external-systems.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+import { Then, When } from "@cucumber/cucumber";
+import { Page } from "@playwright/test";
+import { CustomWorld } from "../support/worlds/world";
+
+const ONE_MINUTE_IN_MS = 60_000;
+
+// Reference to smartdocuments page
+let smartDocumentsPage: Page;
```
Comment (bas-info-nl) [inline src/e2e/step-definitions/external-systems.ts:13]:
```suggestion
let createSdDocumentPopup: Page;
```
Maybe another name is even better?

File: src/e2e/features/4-external-systems.feature [hunk 7]
```diff
@@ -1,8 +1,16 @@
...
     Given "Bob" is logged in to zac
     And Employee "Bob" is on the newly created zaak with status "Intake"
+    
     When Employee "Bob" clicks on Create Document for zaak
-    And Employee "Bob" closes the SmartDocuments tab
-    Then Employee "Bob" should not get an error
+    And Employee "Bob" fills in the create document form
+    And Employee "Bob" submits the form to create the document should see SmartDocuments tab
+    And Employee "Bob" submits the SmartDocuments form
+
```
Comment (bas-info-nl) [inline src/e2e/features/4-external-systems.feature:16]: A scenario should always have a `Then` statement. In this case I'd expect something like:
```gherkin
Then the SmartDocuments form is closed
```

File: src/e2e/features/4-external-systems.feature [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
 
   Scenario: Employee wants to create a new document for a zaak
     Given "Bob" is logged in to zac
     And Employee "Bob" is on the newly created zaak with status "Intake"
+    
     When Employee "Bob" clicks on Create Document for zaak
-    And Employee "Bob" closes the SmartDocuments tab
-    Then Employee "Bob" should not get an error
+    And Employee "Bob" fills in the create document form
+    And Employee "Bob" submits the form to create the document should see SmartDocuments tab
```
Comment (bas-info-nl) [inline src/e2e/features/4-external-systems.feature:14]: The bit "should see SmartDocuments tab" (I'm not clear what that is btw) looks like a `Then` clause to me 🤔

Comment (bas-info-nl) [reply src/e2e/features/4-external-systems.feature:14]: Ah I get it now, that refers to a browser tab 👍 
So, then I should reword this to indicate it is what happens, like so:
```gherkin
And Employee "Bob" submits the form to start the SmartDocuments wizard
```
Underlying code is fine though 😄

File: src/e2e/features/4-external-systems.feature [hunk 9]
```diff
@@ -1,8 +1,16 @@
...
+    And Employee "Bob" submits the SmartDocuments form
```
Comment (bas-info-nl) [inline src/e2e/features/4-external-systems.feature:15]: And this then also refers to the wizard right? Then let's keep that name...
```gherkin
And Employee "Bob" completes the SmartDocuments wizard
```

File: src/e2e/features/4-external-systems.feature
Comment (bas-info-nl) [inline src/e2e/features/4-external-systems.feature:1]: I am not clear why we should refer to "Bob" as an Employee, that is clear from logging in right?

**Review by bas-info-nl — COMMENTED** [review]
Please check the comments, and discuss in slack

File: src/e2e/step-definitions/external-systems.ts (hunk 4 above)
Comment (edgarvonk) [reply src/e2e/step-definitions/external-systems.ts:52]: Yes I think so. Playwright has default timeouts for every await as far as I know.

Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:52]: Playwright has a timeout to 'await'. Though your remark triggered me to find out. And then came to the conclusion that a previous cleanup had cleaned too much code. This is the correct code line:

`await this.expect(submitButton).toBeDisabled();`

There has to be an 'expect' to do the actual assertion.

File: src/e2e/step-definitions/external-systems.ts (hunk 1 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:12]: ✅

File: src/e2e/step-definitions/external-systems.ts (hunk 2 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:34]: ✅

File: src/e2e/step-definitions/external-systems.ts (hunk 3 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:85]: Removed

File: src/e2e/step-definitions/external-systems.ts (hunk 6 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:13]: I called it `smartDocumentsWizardPage` - the 'popup' is just how Playwright calls the event; tmo confusing to put that in the name.

File: src/e2e/step-definitions/external-systems.ts (hunk 5 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts:62]: Fixed (in next push)

File: src/e2e/features/4-external-systems.feature (hunk 8 above)
Comment (infonl-marcel) [reply src/e2e/features/4-external-systems.feature:14]: Totally agreeing here; but the `then` is the failing step that needs to be fixed in either our backend code, or the setup of the callback page or even other. There in no 'then' yet.

Comment (infonl-marcel) [reply src/e2e/features/4-external-systems.feature:14]: And leaving it there calling an empty function seems to me not adding any value;  I removed that whole function btw

File: src/e2e/features/4-external-systems.feature
Comment (infonl-marcel) [reply src/e2e/features/4-external-systems.feature:1]: Just one the personas used; somebody's choice (and humor i guess)

File: src/e2e/step-definitions/external-systems.ts (hunk 4 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts]: Fixed and works; Playwright wait 5 seconds it seems after testing it

File: src/e2e/step-definitions/external-systems.ts (hunk 6 above)
Comment (infonl-marcel) [reply src/e2e/step-definitions/external-systems.ts]: OK to close this @bas-info-nl  ?

File: src/e2e/features/4-external-systems.feature (hunk 7 above)
Comment (infonl-marcel) [reply src/e2e/features/4-external-systems.feature:15]: Totally agreeing here; but the then is the failing step that needs to be fixed in either our backend code, or the setup of the callback page or even other. There in no 'then' yet. And leaving it there calling an empty function seems to me not adding any value; I removed that whole function btw.

File: src/e2e/features/4-external-systems.feature (hunk 8 above)
Comment (infonl-marcel) [reply src/e2e/features/4-external-systems.feature]: Adjusted it; thxz!

**Review by edgarvonk — APPROVED** [review]
noice!
